tools/spy-detector/src/main.rs

```diff
@@ -0,0 +1,262 @@
+                            .inv_witnesstx_received
+                            .fetch_add(1, atomic::Ordering::Relaxed);
+                    } else {
+                        stats
+                            .inv_witnesstx_sent
+                            .fetch_add(1, atomic::Ordering::Relaxed);
+                    }
+                }
+                _ => {} // Ignore other types
+            }
+        }
+    }
+}
+
+fn process_getdata_msg(
+    peer_map: &PeerMap,
+    msg: &net_msg::message::Msg,
+    msg_type: u32,
+    peer_id: u64,
+) {
+    let stats = peer_map
+        .entry(peer_id.to_string())
+        .or_insert_with(PeerStats::default);
+
+    if let net_msg::message::Msg::Inv(inv) = msg {
+        for inv_item in &inv.items {
+            match inv_item.inv_type() {
+                "WitnessTx" => {
+                    if msg_type == 0 {
+                        stats
+                            .getdata_witnesstx_received
+                            .fetch_add(1, atomic::Ordering::Relaxed);
+                    } else {
+                        stats
+                            .getdata_witnesstx_sent
+                            .fetch_add(1, atomic::Ordering::Relaxed);
+                    }
+                }
+                _ => {} // Ignore other types
+            }
+        }
+    }
+}
+
+fn process_tx_msg(peer_map: &PeerMap, msg_type: u32, peer_id: u64) {
+    let stats = peer_map
+        .entry(peer_id.to_string())
+        .or_insert_with(PeerStats::default);
+
+    if msg_type == 0 {
+        stats.tx_received.fetch_add(1, atomic::Ordering::Relaxed);
+    } else {
+        stats.tx_sent.fetch_add(1, atomic::Ordering::Relaxed);
+    }
+}
+
+fn process_connection_event(peer_map: &PeerMap, event: &str) {
+    if event.starts_with("closed") {
+        let peer_id = event.split(' ').nth(1).unwrap_or("");
+        if let Some(stats) = peer_map.remove(peer_id) {
+            println!("Connection closed for peer: {}", peer_id);
+            print_peer_stats(peer_id, &stats.1);
+        }
+    }
+}
+
+fn print_peer_stats(peer_addr: &str, stats: &PeerStats) {
+    println!(
+        "[{}] Peer ID {} stats:\n  INV TX sent: {:?}\n  INV TX received: {:?}\n  INV WTX received: {:?}\n INV WitnessTX received: {:?}\n GetDATA WitnessTX received: {:?}\n  GETDATA WitnessTX received: {:?}\n  Tx sent: {:?}\n  Tx received: {:?}\n Last Activity: {:?}",
+        chrono::Local::now().format("%Y-%m-%d %H:%M:%S"),
+        peer_addr,
+        stats.inv_tx_sent,
+        stats.inv_tx_received,
+        stats.inv_wtx_received,
+        stats.inv_witnesstx_received,
+        stats.getdata_witnesstx_sent,
+        stats.getdata_witnesstx_received,
+        stats.tx_sent,
+        stats.tx_received,
+        stats.last_activity
+    );
+}
+
+fn display_all_stats(peer_map: &PeerMap) {
+    let map = peer_map;
+    println!(
+        "\n===== Stats for all peers ({}): =====",
+        chrono::Local::now().format("%Y-%m-%d %H:%M:%S")
```

Comment:
I don't hink we need the timestamp as we usually only care about the most recent one.

Reply:
Okay, removing timestamps for now!!